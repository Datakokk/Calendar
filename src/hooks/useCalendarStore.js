import { useDispatch, useSelector } from "react-redux"
import Swal from "sweetalert2";
import { calendarApi } from "../api";
import { convertEventsToDateEvents } from "../helpers";
import { onAddNewEvent, onDeletEvent, onLoadEvents, onSetActiveEvent, onUpdateEvent } from "../store/calendar/calendarSlice";

export const useCalendarStore = () => {

    const dispatch = useDispatch();

    const { events, activeEvent } = useSelector( store => store.calendar );
    const { user } = useSelector( store => store.auth );
    
    const setActiveEvent = ( calendarEvent ) => {

        dispatch( onSetActiveEvent( calendarEvent ) );
    };

    const starSavingEvent = async( calendarEvent ) => {
        try {
            
            // Everything it is ok
            if( calendarEvent.id){
                //Update
                await calendarApi.put(`/events/update/${calendarEvent.id}`, calendarEvent);
                dispatch( onUpdateEvent( {...calendarEvent, user} ))
                return;
            };

            // Create
            const { data } = await calendarApi.post('/events/create', calendarEvent );
            dispatch( onAddNewEvent({...calendarEvent, id: data.event.id, user }));
        } catch (error) {
            console.log(error);
            Swal.fire('Failed to save', error.response.data.msg, 'error');
        }
    };

    const starDeletingEvent = async() => {
        try {
            
            await calendarApi.delete(`/events/delete/${activeEvent?.id}`);
            dispatch( onDeletEvent() );

        } catch (error) {
            console.log(error);
            Swal.fire('Failed to delete', error.response.data.msg, 'error');
        }
    }

    const startLoadingEvents = async () => {

        try {
            
            const { data } = await calendarApi.get('/events');
            const events = convertEventsToDateEvents( data.event )
            dispatch( onLoadEvents( events ));
            
        } catch (error) {
            console.log('Failed to load events')
            console.log(error)
        }
    }
    return {
        // Properties
        events,
        activeEvent,
        hasEventSelected: !!activeEvent,

        //* Methods
        setActiveEvent, 
        starDeletingEvent,
        starSavingEvent,
        startLoadingEvents
    }
}
