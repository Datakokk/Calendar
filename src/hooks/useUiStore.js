import { useDispatch, useSelector } from "react-redux"
import { onCloseDateModal, onopenDateModal } from "../store/ui/uiSlice";

export const useUiStore = () => {

    const dispatch = useDispatch();
   
    const { isDataModalOpen } = useSelector( store => store.ui );

    const openDateModal = () => {
        dispatch( onopenDateModal() );
    }

    const closeDateModal = () => {
        dispatch( onCloseDateModal() );
    }

    const toggleDateModal = () => {
        isDataModalOpen 
        ? dispatch( onopenDateModal )
        : dispatch( onCloseDateModal )
    }
    return {
        // Properties
        isDataModalOpen,

        //* Methods
        openDateModal,
        closeDateModal,
        toggleDateModal
    }
}
