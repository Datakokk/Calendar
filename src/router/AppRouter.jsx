import { Navigate, Route, Routes } from "react-router-dom";
import { LoginPage } from "../auth";
import { Calendarpage } from "../calendar";

export const AppRouter = () => {

    const authStatus = 'authenticated'; //'not-authenticated' 'authenticated'

  return (
    <Routes>
        
        {
            (authStatus === 'not-authenticated')
            ?<Route path="/auth/*" element={ <LoginPage/> }/>
            :<Route path="/*" element={ <Calendarpage /> } />
        }

        <Route path="/*" element={ <Navigate to="/auth/login" /> } />
    </Routes>
  )
}
