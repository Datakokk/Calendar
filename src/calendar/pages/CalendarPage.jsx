import { useEffect, useState } from 'react'
import { Calendar } from 'react-big-calendar'
import 'react-big-calendar/lib/css/react-big-calendar.css'

import { Navbar, CalendarEvent, CalendarModal, FabAddNew, FabDelete } from '..'

import { localizer, getMessagesNO } from '../../helpers'
import { useAuthStore, useCalendarStore, useUiStore } from '../../hooks'

export const CalendarPage = () => {

  const { user } = useAuthStore();

  const { events, setActiveEvent, startLoadingEvents } = useCalendarStore();

  const { openDateModal } = useUiStore();

  const [lastView, setLastView] = useState(localStorage.getItem('lastView') || 'month')

  const eventStyleGetter = (event, start, end, isSelected) => {

    const isMyEvent = ( user.uid === event.user._id || user.uid === event.user.uid );
    const style = {
      backgroundColor: isMyEvent ? '#347Cf7': '#2ECC71',
      borderRadius: '0px',
      opacity: 0.8,
      color: '#fff'
    }

    return {
      style
    }
  }

  const onDoubleClick = (event) => {
    // console.log({ DoubleClickEvent: event })
    openDateModal();
  }
  const onSelect = (event) => {
    // console.log({ SelectEvent: event })
    setActiveEvent( event )
  }
  const onViewChanged = (event) => {
    localStorage.setItem('lastView', event);
  }

  useEffect(() => {
    startLoadingEvents();
  }, [])
  
  return (
    <>
      <Navbar />

      <Calendar
        culture='nb'
        localizer={localizer}
        events={events}
        defaultView={lastView}
        startAccessor="start"
        endAccessor="end"
        style={{ height: 'calc( 100vh - 80px )' }}
        messages={getMessagesNO()}
        eventPropGetter={eventStyleGetter}
        components={{
          event: CalendarEvent
        }}
        onDoubleClickEvent={onDoubleClick}
        onSelectEvent={onSelect}
        onView={onViewChanged}
      />

      <CalendarModal />
      <FabAddNew />
      <FabDelete />
    </>
  )
}
