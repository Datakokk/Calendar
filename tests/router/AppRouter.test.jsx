import { render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom";
import { CalendarPage } from "../../src/calendar/pages/CalendarPage";
import { useAuthStore } from "../../src/hooks/useAuthStore"
import { AppRouter } from "../../src/router/AppRouter"

jest.mock('../../src/hooks/useAuthStore');
jest.mock('../../src/calendar/pages/CalendarPage', () => ({
    CalendarPage: () => <h2>CalendarPage Test</h2>
}));

describe('Test on <AppRouter />', () => {

    const mockCheckAuthToken = jest.fn();

    test('should the screen load and call checkAuthToken', () => {

        useAuthStore.mockReturnValue({
            status: 'checking',
            checkAuthToken: mockCheckAuthToken

        })

        render(<AppRouter />)

        expect(screen.getByText('Loading...')).toBeTruthy();
        expect(mockCheckAuthToken).toHaveBeenCalled();
    });

    test('should display  login if  user is not-authenticated', () => {

        useAuthStore.mockReturnValue({
            status: 'not-authenticated',
            checkAuthToken: mockCheckAuthToken

        })

        const { container } = render(
            <MemoryRouter initialEntries={['/auth2/something/new_thing']}>
                <AppRouter />
            </MemoryRouter>
        );

        expect('Logg inn').toBeTruthy();
        expect(container).toMatchSnapshot()
    });

    test('should display the calendar if we are authenticated', () => {

        useAuthStore.mockReturnValue({
            status: 'authenticated',
            checkAuthToken: mockCheckAuthToken

        })

        render(
            <MemoryRouter >
                <AppRouter />
            </MemoryRouter>
        );
        expect( screen.getByText('CalendarPage Test')).toBeTruthy()
        
    });
})