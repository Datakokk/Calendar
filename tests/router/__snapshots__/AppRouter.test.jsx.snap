// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <AppRouter /> should display  login if  user is not-authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Logg inn
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="E-post"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Passord"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Melde deg på
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Navn"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="E-post"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Passord"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Gjenta passord"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Opprett konto"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test on <AppRouter /> should display the calendar if we are authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Logg inn
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="E-post"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Passord"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Melde deg på
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Navn"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="E-post"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Passord"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Gjenta passord"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Opprett konto"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
